### PR TITLE
[how the fuck do i fix this bug please to help] overhaul: ling verbs to action buttons

### PR DESCRIPTION
**Pull request by PKPenguin321**
:cl: PKPenguin321, Armhulen, epochayur
refactor: Changeling powers are now action buttons rather than verbs.
imageadd: Thanks to epochayur for creating and compiling sprites for every single changeling action button!
imageadd: Thanks to Armhulen for the changeling button background sprite!
/:cl:

Still very messy at this stage but fully functional and compiles. Was even able to playtest it and it works as intended.

# **Cant be finished because theres a really stupid bug where if you change bodies (cloning, last resort, monkeying) you lose all your powers. Cant figure out a fix please HELP**

Things that still need doing:
- [x]  Clean up comments and old code
- [x] Move the rest of the verbs to buttons (the systems are all in place but currently I've only moved transformation sting and mimic voice for testing reasons). Probably the most tedious step
- [x] Get sprites for every power that needs a button (passives won't need a button, stings can just use the sting icon). Not technically required but much needed polish.
- [ ] This will assuredly need playtesting, maybe live test merges. (It got one once before and found the bug described above)

### CODE REVIEW

**Comment by PKPenguin321** [comment]
If you know who made the bg_changeling icon please let me know so I can credit them. I stole it from arm's thingling PR. Will also credit anybody who gives sprites

**Comment by Armhulen** [comment]
It was me muahahahaa

**Comment by Armhulen** [comment]
Thanks for doing this for me btw, I was about to do the exact same deal

**Comment by uraniummeltdown** [comment]
based pk

**Comment by Kromgar** [comment]
Will this be a seperate action bar button like bloodcult?

**Comment by PKPenguin321** [comment]
I wasn't aware a second bar was possible, if you can point me to relevant code and it's not too hard to implement I can see what I can do

**Comment by Kromgar** [comment]
@PKPenguin321 Cultcode the action buttons for spells/runes.  @Robustin  might be able to help? Also juggernauts, harvesters, and i think artificers use it too

**Comment by ma44** [comment]
@PKPenguin321 When a action bar gets too long, additional action buttons will be added beneath it.

**Comment by PKPenguin321** [comment]
I know but I don't think that's what he means

**Comment by Kromgar** [comment]
Cult has an additional  action bar seperate from the one in the top left

**Comment by PKPenguin321** [comment]
this won't compile until every power is transferred over, right now i can compile locally and test by keeping the old changling_powers.dm in the .dme though (kind of a hack but it's the only way to test before i'm absolutely finished without getting errors on compile)

**Comment by MrDoomBringer** [comment]
@Kromgar, the cult icons to appear at the bottom of the screen, but I believe that's an error, as if you reset button positions they all line up nicely at the top left.

**Comment by Potato-Masher** [comment]
The blood cult button locations are intentional.

**Comment by Potato-Masher** [comment]
Also:

>inb4 having three rows of action buttons because of all the damn abilities lings have at round start

**Comment by PKPenguin321** [comment]
You can drag buttons around to reposition them you know

**Comment by epochayur** [comment]
pkp you dumb slut, reply to me on discord

are the powers listed under '/code/modules/antagonists/changeling/powers/' the only ones that need icons? if so I'll sprite those up

**Comment by PKPenguin321** [comment]
Yeah but a few of those files have more than one power in them. [The list on the wiki's page for changelings is more accurate](https://tgstation13.org/wiki/Changeling), but that one is missing Hivemind Upload/Download and Absorb.

**Comment by Armhulen** [comment]
@PKPenguin321

Incase you missed it, I made the background sprite

**Comment by PKPenguin321** [comment]
Yeah I saw, just havent updated the 🆑 yet

**Comment by PKPenguin321** [comment]
Up next: Fixing compile errors

**Comment by PKPenguin321** [comment]
Only thing that needs changing from a code perspective now is probably the cellular emporium formatting and stylizing, which is a bit fucked up. Also will definitely need a test merge since I'm not 100% confident it'll be bug free.

**Comment by PKPenguin321** [comment]
well the cellular emporium works now but upon testing i now see that some abilities just don't like getting removed on respec which is a pretty big problem. will test with it some more and keep working...

**Comment by PKPenguin321** [comment]
have u ever felt retarded

**Comment by PKPenguin321** [comment]
ok. code side is basically done, aside from needing review (and having to add lines for every button that add icons once i get real sprites).
once i get sprites it will be much much prettier, and after that it should probably be tested, but once that's finished it should be good to merge! phew

File: code/modules/antagonists/changeling/changeling.dm
Comment (MoreRobustThanYou) [inline code/modules/antagonists/changeling/changeling.dm]: FALSE

File: code/modules/antagonists/changeling/changeling.dm
Comment (MoreRobustThanYou) [inline code/modules/antagonists/changeling/changeling.dm]: i don't think the comment is needed

File: code/modules/antagonists/changeling/changeling.dm
Comment (MoreRobustThanYou) [inline code/modules/antagonists/changeling/changeling.dm]: comment not needed

File: code/modules/antagonists/changeling/changeling.dm
Comment (MoreRobustThanYou) [inline code/modules/antagonists/changeling/changeling.dm]: why is your code full of obvious comments

File: code/modules/antagonists/changeling/changeling.dm
Comment (MoreRobustThanYou) [inline code/modules/antagonists/changeling/changeling.dm]: stop with the dumb comments

**Review by ghost — APPROVED** [review]
Looks good to me, just cut out the shitty comments

File: code/modules/antagonists/changeling/changeling.dm
Comment (PKPenguin321) [reply code/modules/antagonists/changeling/changeling.dm]: my b a lot of these were for self reference to get a grasp on existing ling code and some were TODOs that i forgot to put a TODO on

**Comment by PKPenguin321** [comment]
Hm. Just noticed that I accidentally cut out the powers listing their cost. Will readd at the same time I get sprites since I'll need to comb through every button again

File: code/modules/antagonists/changeling/powers/hivemind.dm
Comment (lzimann) [inline code/modules/antagonists/changeling/powers/hivemind.dm]: typeless loop and only assigned the var if the next check is true

File: code/modules/antagonists/changeling/changeling.dm
Comment (lzimann) [inline code/modules/antagonists/changeling/changeling.dm]: typeless loop

File: code/modules/antagonists/changeling/changeling.dm
Comment (lzimann) [inline code/modules/antagonists/changeling/changeling.dm]: keep this `TRUE`, no real reason to return the instance here

**Review by lzimann — CHANGES_REQUESTED** [review]
this looks fine except a few minor things

File: code/modules/antagonists/changeling/changeling.dm
Comment (PKPenguin321) [reply code/modules/antagonists/changeling/changeling.dm]: What does this mean

Comment (lzimann) [reply code/modules/antagonists/changeling/changeling.dm]: basically:
```
for(var/power in purchasedpowers)
    var/datum/action/changeling/P = power
    ....
```

everything in this list is already this type, so you don't need to do a type check(which is what this kind of loop does)

Comment (PKPenguin321) [reply code/modules/antagonists/changeling/changeling.dm]: appreciate it

Comment (MoreRobustThanYou) [reply code/modules/antagonists/changeling/changeling.dm]: Microoptimizations

**Review by lzimann — DISMISSED** [review]

**Comment by epochayur** [comment]
https://mega.nz/#!I48XFCZT!hi94DDSlAMatPmNGLZRR5rx7ljQKcu64KGX8ZyGhAIg wew okay, here are the sprites, apologies again for taking so long

These will definitely need to be renamed, and I realize the quality of the original sprites may be a little shoddy, but it's better than nothing until someone can come along and change them. A decent amount of these are lifted from the changeling items in items_and_weapons.dmi . Ping me if you need any changes

**Comment by PKPenguin321** [comment]
Thanks a ton. Only one I noticed was missing is Biodegrade, the one that lets you melt cuffs

**Comment by PKPenguin321** [comment]
hold that thought just found a bug that makes reviving stasis not work at all

**Comment by epochayur** [comment]
https://mega.nz/#!B9N3gCqZ!UDiujVxLaRDIdGdUUcI_LvScjpz6yuxWRhxlKepBslY new link, includes biodegrade and hivemind_link, just use whichever for hivemind you think looks better

**Comment by PKPenguin321** [comment]
sprites are excellento now. thanks so much again

and if i wasnt clear with my last comment please hold off on testmerging this until i fix this bug, not sure how i didn't catch it before

File: code/modules/antagonists/changeling/powers/fakedeath.dm
Comment (PKPenguin321) [inline code/modules/antagonists/changeling/powers/fakedeath.dm]: the name, desc, and button_icon_state changes aren't reflected on the button itself. everything else works but this. not sure why.

Comment (YPOQ) [reply code/modules/antagonists/changeling/powers/fakedeath.dm]: These vars are on the action datum, call UpdateButtonIcon() to copy them over to the button.

**Comment by PKPenguin321** [comment]
wait what it failed

File: code/modules/antagonists/changeling/powers/adrenaline.dm
Comment (PKPenguin321) [inline code/modules/antagonists/changeling/powers/adrenaline.dm]: Bad TODO

File: code/modules/antagonists/changeling/powers/fakedeath.dm
Comment (PKPenguin321) [inline code/modules/antagonists/changeling/powers/fakedeath.dm]: Forgot to update desc and icon here as well

File: code/modules/antagonists/changeling/powers/fakedeath.dm
Comment (PKPenguin321) [inline code/modules/antagonists/changeling/powers/fakedeath.dm]: now that I think about it this has other checks meaning this can probably always be true

**Comment by PKPenguin321** [comment]
BUGS FOUND:
- Transforming to a monkey does not properly grant the human form button and removes all of your buttons. Rehumanizing does not give them back
- Dying and being cloned (which puts you into a new body) causes you to lose all of your buttons

**Comment by PKPenguin321** [comment]
Sorry I haven't touched this in so long, been a bit low on free time lately. Going to look at this and see if there's anything I can do to fix it up.

As far as I could tell, old code for verbs had the verbs list being updated on Stat() which seemed like an awful hack so I removed it, and I suspect that this has something to do with why powers aren't being carried over correctly on transform. My plan right now is to I think make a proc that re-grants all purchased powers which I can run whenever the mind with the antag datum enters a new body, but I'm not sure if there's any proc that's ran when a mind enters a new body.

**Comment by PKPenguin321** [comment]
Also note to self I should test what happens if a non-ling mindswaps into a ling

**Review by ghost — CHANGES_REQUESTED** [review]
I'd personally prefer a /datum/changeling_action, with them auto-generating action datums based on vars, but this is fine, too.

**Comment by PKPenguin321** [comment]
my last commit will most likely fail travis. i dont understand the warning the compiler is throwing but playtesting it shows that my solution doesnt even work except for the cellular emporium button and i have no fuckin clue why that might be.
i just dont get it

**Comment by optimumtact** [comment]
travis

**Comment by PKPenguin321** [comment]
no merge conflicts on pull?

also i cant fix this bug i dont know how and im out of steam on reading all the fuckign transformation code to figure out how.

**Comment by PKPenguin321** [comment]
i dont know what the compiler warning on my attempted fix means either. either way the fix didnt work so i guess i can just remove it so at least travis passes.

File: code/modules/antagonists/changeling/changeling.dm
Comment (optimumtact) [inline code/modules/antagonists/changeling/changeling.dm]: you know this is required code right?

Comment (PKPenguin321) [reply code/modules/antagonists/changeling/changeling.dm]: Yes it kills travis see the comments below it
It does nothing in testing also i wrote it so yes i know what it is

**Comment by SpaceManiac** [comment]
You wrote:
```dm
	for(var/path in purchasedpowers)
		var/datum/action/changeling/S = new path()
			S.Grant(owner.current)
```
You probably meant:
```dm
	for(var/path in purchasedpowers)
		var/datum/action/changeling/S = new path()
		S.Grant(owner.current)
```

**Comment by PKPenguin321** [comment]
holy heck you might be right

**Comment by PKPenguin321** [comment]
ok so it still didnt work, only the cellular emporium button is granted by regain_powers and the rest of the buttons are still lost. at least it passes travis now without commenting it out. will try putting it on mob login.

**Comment by PKPenguin321** [comment]
issue persists even when the proc is ran on mob login 😠 😢

**Comment by PKPenguin321** [comment]
why the fuck my code dont work
why

**Comment by KorPhaeron** [comment]
I was excited for this PR but it is pretty clearly dead, please finish it when you get the chance

**Comment by PKPenguin321** [comment]
basically action buttons seem to poof out of existence and DO NOT want to be regranted if you change mob for some reason. i suspect it's a bug with action buttons but am not about to read all the action button code to figure it out, just too exhausted with it. rip

**Pull request closed without merging**
